linkerd/proxy/transport/src/listen.rs

```diff
@@ -88,7 +92,16 @@ where
             let tcp = res.map_err(AcceptError)?;
             super::set_nodelay_or_warn(&tcp);
             let tcp = super::set_keepalive_or_warn(tcp, keepalive).map_err(KeepaliveError)?;
-            let client = Remote(ClientAddr(tcp.peer_addr().map_err(PeerAddrError)?));
+            let mut client_addr = tcp.peer_addr().map_err(PeerAddrError)?;
+
+            // unwrap IPv4-mapped IPv6 address
+            if let SocketAddr::V6(ipv6) = client_addr {
+                if let Some(ipv4) = ipv6.ip().to_ipv4_mapped() {
+                    client_addr = SocketAddr::V4(SocketAddrV4::new(ipv4, client_addr.port()));
+                }
+            }
+
+            let client = Remote(ClientAddr(client_addr));
```

Comment:
golf:
* a dedicated function let's us avoid `mut` with an early return
* Into::into conversion is much less verbose

```suggestion

            fn ipv4_mapped(orig: SocketAddr) -> SocketAddr {    
                if let SocketAddr::V6(v6) = orig {
                    if let Some(ip) = v6.ip().to_ipv4_mapped() {
                        return (ip, orig.port()).into();
                    }
                }
                orig
            }
            
            let client_addr = tcp.peer_addr().map_err(PeerAddrError)?;
            let client = Remote(ClientAddr(ipv4_mapped(client_addr)));
```

Reply:
Ah nice!